flowmachine/flowmachine/features/location/majority_location.py

```diff
@@ -0,0 +1,56 @@
+from typing import List
+
+from flowmachine.core import Query
+from flowmachine.core.mixins import GeoDataMixin
+from flowmachine.features.utilities.subscriber_locations import BaseLocation
+
+
+class MajorityLocation(BaseLocation, Query):
+    def __init__(
+        self,
+        subscriber_location_weights: Query,
+        weight_column,
+        include_unlocatable=False,
+    ):
+        if weight_column not in subscriber_location_weights.column_names:
+            raise ValueError("weight_column must exist in subscriber_subset")
+        self.subscriber_location_weights = subscriber_location_weights
+        self.weight_column = weight_column
+        self.include_unlocatable = include_unlocatable
+        self.spatial_unit = subscriber_location_weights.spatial_unit
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        # This should be a reduction, so no new columns
+        return [
+            "subscriber"
+        ] + self.subscriber_location_weights.spatial_unit.location_id_columns
+
+    def _make_query(self):
+        loc_id = ",".join(
+            self.subscriber_location_weights.spatial_unit.location_id_columns
+        )
+        sql = f"""
+WITH subscriber_subset AS (
+    {self.subscriber_location_weights.get_query()}
+), summed_weights AS (
+    SELECT subscriber, sum({self.weight_column}) AS total_weight
+    FROM subscriber_subset
+    GROUP BY subscriber
+), seen_subs AS (
+    SELECT subscriber, {loc_id}
+    FROM summed_weights JOIN subscriber_subset USING(subscriber)
+    WHERE {self.weight_column} > total_weight/2.0
```

Comment:
A thought - I think that this would still work if `weight_column` is a float, but I'm not sure if there's an issue if it contains zeros.

Reply:
I think it should be fine. If `weight_column=0` then `weight_column <= total_weight/2.0` (provided there are no negative weights), so that row won't be included.

Could get some unexpected behaviour if there are negative weights (including getting multiple output rows per subscriber), but that would be a strange way to use this query. I'm not sure it's worth explicitly checking that all weights are >=0, though.

Reply:
Can't check the weights are all >=0 ahead of time without executing the sql, right? You could coerce them to be so, or equally add a warning to the docstring?

Reply:
If there's a single obvious choice for how to coerce negative weights to be >=0 then I think I'd be in favour of that (along with a warning in the docstring). Setting negative weights equal to zero (effectively the same as just ignoring them) seems the most appropriate option, although using the absolute value is possibly an alternative.

Reply:
Easy route is to filter out using `WHERE weight >= 0`?

Reply:
Sorry, read the discussion properly now. I do think that an explicit check is worth it, as we're already visiting each row as part of the summing-up - I don't think that this would add much to the execution time.

Reply:
I'm not sure I'm in favour of that. It's not so much the extra cost of the check, but the fact that the check wouldn't happen until you try to get the result. You could create the query, submit it, wait for it to get to the top of the queue, and only then find out it's not valid. I think filtering out negative weights is preferable, provided there's a warning in the docstring.

Reply:
Agree with that - filter in place, and docstring.